aptos-move/aptos-vm/src/validator_txns/jwk.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    aptos_vm::get_or_vm_startup_failure,
+    errors::expect_only_successful_execution,
+    move_vm_ext::{AptosMoveResolver, SessionId},
+    system_module_names::{JWKS_MODULE, UPSERT_INTO_OBSERVED_JWKS},
+    validator_txns::jwk::{
+        ExecutionFailure::{Expected, Unexpected},
+        ExpectedFailure::{
+            IncorrectVersion, MissingResourceObservedJWKs, MissingResourceValidatorSet,
+            MultiSigVerificationFailed, NotEnoughVotingPower,
+        },
+    },
+    AptosVM,
+};
+use aptos_bitvec::BitVec;
+use aptos_types::{
+    aggregate_signature::AggregateSignature,
+    fee_statement::FeeStatement,
+    jwks,
+    jwks::{Issuer, ObservedJWKs, ProviderJWKs, QuorumCertifiedUpdate},
+    move_utils::as_move_value::AsMoveValue,
+    on_chain_config::{OnChainConfig, ValidatorSet},
+    transaction::{ExecutionStatus, TransactionStatus},
+    validator_verifier::ValidatorVerifier,
+};
+use aptos_vm_logging::log_schema::AdapterLogSchema;
+use aptos_vm_types::output::VMOutput;
+use move_core_types::{
+    account_address::AccountAddress,
+    value::{serialize_values, MoveValue},
+    vm_status::{AbortLocation, StatusCode, VMStatus},
+};
+use move_vm_types::gas::UnmeteredGasMeter;
+use std::collections::HashMap;
+
+enum ExpectedFailure {
+    // Move equivalent: `errors::invalid_argument(*)`
+    IncorrectVersion = 0x010103,
+    MultiSigVerificationFailed = 0x010104,
+    NotEnoughVotingPower = 0x010105,
+
+    // Move equivalent: `errors::invalid_state(*)`
+    MissingResourceValidatorSet = 0x30101,
+    MissingResourceObservedJWKs = 0x30102,
+}
+
+enum ExecutionFailure {
+    Expected(ExpectedFailure),
+    Unexpected(VMStatus),
+}
+
+impl AptosVM {
+    pub(crate) fn process_jwk_update(
+        &self,
+        resolver: &impl AptosMoveResolver,
+        log_context: &AdapterLogSchema,
+        session_id: SessionId,
+        update: jwks::QuorumCertifiedUpdate,
+    ) -> Result<(VMStatus, VMOutput), VMStatus> {
+        match self.process_jwk_update_inner(resolver, log_context, session_id, update) {
+            Ok((vm_status, vm_output)) => Ok((vm_status, vm_output)),
+            Err(Expected(failure)) => {
+                // Pretend we are inside Move, and expected failures are like Move aborts.
```

Comment:
why do we need to simulate move abort? normal move abort is not discard too?

Reply:
We don't need to. But I had the impression that In the future we may want to move the checks into move.

Discarding or not is a separate question? And for vtxn we want to discard if the checks failed?